### PR TITLE
Clean up copy specs

### PR DESCRIPTION
**Pull request by lptr**
- Add `@Override` annotations where necessary
- Add `@Nullable` and `@NonNullApi` annotations where necessary
- Some small optimizations
- Fix IntelliJ warnings
- Remove some dead code

This is about salvaging as much as possible from #4655.

### CODE REVIEW

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=11552054

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=11552853

**Comment by lptr** [comment]
New build: https://builds.gradle.org/viewQueued.html?itemId=11556021

**Comment by lptr** [comment]
@wolfs since all tests pass I'm going to merge this now. Feel free to review post-merge, though.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java [hunk 1]
```diff
@@ -55,30 +59,28 @@
 public class DefaultCopySpec implements CopySpecInternal {
     private static final NotationParser<Object, String> PATH_NOTATION_PARSER = PathNotationConverter.parser();
     protected final FileResolver fileResolver;
-    private final Set<Object> sourcePaths;
+    private final Set<Object> sourcePaths = new LinkedHashSet<Object>();
     private Object destDir;
     private final PatternSet patternSet;
-    private final List<CopySpecInternal> childSpecs;
-    private final List<CopySpecInternal> childSpecsInAdditionOrder;
+    private final List<CopySpecInternal> childSpecs = new LinkedList<CopySpecInternal>();
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java]: Why did you change `childSpecs`, `childSpecsInAdditionOrder` and `copyActions` from `ArrayList` to `LinkedList`?

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java [hunk 2]
```diff
@@ -208,20 +216,24 @@ public void addChildSpecListener(CopySpecListener copySpecListener) {
         this.listeners.add(copySpecListener);
     }
 
+    @VisibleForTesting
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java]: Public visibility for testing seems wrong...

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java [hunk 3]
```diff
@@ -445,44 +491,39 @@ public void appendCachingSafeCopyAction(Action<? super FileCopyDetails> action)
...
     }
 
+    @Override
     public String getFilteringCharset() {
         return buildRootResolver().getFilteringCharset();
     }
 
-    public void setFilteringCharset(String charset) {
+    @Override
+    public void setFilteringCharset(@Nullable String charset) {
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java]: Can't we just remove the `@Nullable` here? We are throwing an exception when null is passed and the getter can't return null, either.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java]: Because there are typically no child specs or a single or a couple child specs. `ArrayList` optimizes for the no child spec case, but goes on to allocate 10 entries as soon as you add something. I expect a `LinkedList` is more efficient for these cases, e.g. the single child most root specs have.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java (hunk 2 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java]: Having unit tests accessing this from other packages seems wrong, but at least now it's marked for testing. It's not making things worse.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java (hunk 3 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java]: That's a good point.

**Comment by lptr** [comment]
@wolfs thanks for the review, please take a look at https://github.com/gradle/gradle/pull/4721
